src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionService.cs

```diff
@@ -0,0 +1,695 @@
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    return new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0);
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item, filterText, data.InitialTrigger.Reason, data.UpdateTrigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Deletion ||
+                        data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                return HandleAllItemsFilteredOut(data.InitialTrigger.Reason, data.SelectedFilters, selectedFilters, mustSetSelection);
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion
+            // ourselves.
+            if (data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Deletion)
+            {
+                return HandleDeletionTrigger(data.InitialSortedList, data.InitialTrigger.Reason, data.SelectedFilters, data.UpdateTrigger.Reason, filterText, initialListOfItemsToBeIncluded, mustSetSelection);
+            }
+
+            var caretPoint = session.TextView.GetCaretPoint(data.Snapshot.TextBuffer);
```

Comment:
Is this safe to call? Is there some thread races here if we're remapping buffers at the wrong time?

Reply:
this is not safe to call. This method is invoked in a background thread and user has potentially (and it is very likely) typed more and therefore moved the caret.
Myself and the editor team are open to talking about this because we would like to replace ITextView with something smaller and living in Text\Logic layer, as opposed to the UI layer.